src/conditions/condition-expr.ts

```diff
@@ -85,9 +85,10 @@ export class ConditionExpression {
   }
 
   public buildContext(
-    provider: ethers.providers.Web3Provider
+    provider: ethers.providers.Provider,
+    signer: ethers.Signer
```

Comment:
Does `buildContext` get called if there are no context variables in the condition, eg. blocktime condition?

What if there are only user-defined context variables since they don't need anything signed?

Basically, I'm wondering if we properly handle signer being undefined if no signing is needed for the context...either because the condition does need it (in which case buildContext may not even be called), or if there are only user-defined custom context variables in which case there is nothing to be signed.

Reply:
We could check `ConditionExpression` for the presence of context parameters and avoid passing `signer` in if it doesn't contain one. But we still need to make `signer` available in the user-facing API in case we need it (https://github.com/nucypher/nucypher-ts/blob/sep-signer-provider/src/characters/cbd-recipient.ts#L70)

I was considering this and other cases where the signer is optional. Which one of these approaches do we prefer?
1. Make the signer optional and throw an error if it turns out we actually need it
2. Split API into two user flows and make it required in one of them and remove it from the other.
I'm leaning towards the first option. I'm assuming that API users can detect/handle these cases correctly.
If we'll agree on 1), I will simply add optionality and relevant errors in this PR.

Reply:
I tentatively implemented these changes.

Reply:
I think 1) is reasonable. 

> I'm assuming that API users can detect/handle these cases correctly.

Is there a public API/utility function that can detect whether a signer is needed based on the condition?

Reply:
I think it's the `ThresholdDecrypter.retrieveAndDecrypt`, which will be exposed by `taco.decrypt`

Reply:
I guess I'm wondering is if I'm an API user, 
- if a signer is needed and one is not provided I get prompted somehow 
OR 
- do I need to inspect the condition myself to determine if `:userAddress` is in it and therefore I know I need to provide a signer
OR
- is there a function on a condition object, or the decrypter, or some other utility that provides me with a bool result as to whether I need the signer or not based on the condition
OR
...

TL;DR 😅  If signer is optional, how does the API user determine whether to pass it or not?

Reply:
> if a signer is needed and one is not provided I get prompted somehow

There are only two ways that the user can learn that:
- From a function signature (it's "marked" as required, throws an error if missing during TS compilation etc.)
- From run-time errors - The parameter is marked as optional, but there is some dependency on it being present under certain conditions (context parameter is present or not). And because of this hidden dependency, we have to throw an exception in runtime instead

> do I need to inspect the condition myself to determine if :userAddress is in it and therefore I know I need to provide a signer

If you want to avoid having errors thrown in runtime, then yes. 

> is there a function on a condition object, or the decrypter, or some other utility that provides me with a bool result as to whether I need the signer or not based on the condition

No, but we can add it.

Reply:
If `signer` is optional then I think providing such a facility would be good. cc @KPrasch , @cygnusv in case they have a different view.

Reply:
I've added a `ConditionExpression.contextRequiresSigner` helper function and related utilities. This is something end-users may use to check if they need to use a signer for their conditions (to build the context). We should document in an advanced usage guide.